tracing/test-agent-shutdown.sh

```diff
@@ -0,0 +1,1341 @@
+{
+	local cmd="check_local_agent_alive"
+
+	info "Waiting for agent process to start"
+
+	waitForProcess \
+		"$wait_time_secs" \
+		"$sleep_time_secs" \
+		"$cmd"
+
+	info "Kata agent process running"
+}
+
+# Create a Kata Container that blocks "forever"
+start_agent_in_kata_vm()
+{
+	local log_file="${1:-}"
+	[ -z "$log_file" ] && die "need agent log file"
+
+	local snapshotter=""
+	local ret
+
+	# Allow containerd to run on a ZFS root filesystem
+	{ zfs list &>/dev/null; ret=$?; } || true
+	[ "$ret" = 0 ] && snapshotter='zfs'
+
+	# Ensure the container blocks forever
+	local cmd='tail -f /dev/null'
+
+	run_cmd \
+		"tmux new-session \
+		-d \
+		-s \"$KATA_TMUX_VM_SESSION\" \
+		\"sudo ctr run \
+			--snapshotter '$snapshotter' \
+			--runtime '${RUNTIME}' \
+			--rm \
+			-t '${CTR_DOCKER_IMAGE}' \
+			'$container_id' \
+			$cmd\""
+
+	[ "$dry_run" = 'false' ] && \
+		wait_for_kata_vm_agent_to_start "$container_id" || true
+}
+
+start_agent()
+{
+	local agent_test_type="${1:-}"
+	[ -z "$agent_test_type" ] && die "need agent test type"
+
+	local log_file="${2:-}"
+	[ -z "$log_file" ] && die "need agent log file"
+
+	case "$agent_test_type" in
+		'local') start_local_agent "$log_file" ;;
+		'vm') start_agent_in_kata_vm "$log_file" ;;
+		*) die "invalid agent test type: '$agent_test_type'" ;;
+	esac
+
+	true
+}
+
+run_agent_ctl()
+{
+	local server_addr="${1:-}"
+
+	shift
+
+	local cmds="${*:-}"
+
+	[ -n "$server_addr" ] || die "need agent ttRPC server address"
+	[ -n "$cmds" ] || die "need commands for agent control tool"
+
+	# Note that we don't use the same logic for checking the control tool
+	# since that uses a different build profile compared to the agent.
+	command -v "$agent_ctl_binary_name" &>/dev/null || \
+		(cd "$agent_ctl_dir" && cargo install --path .)
+
+	local agent_ctl_path
+	agent_ctl_path=$(command -v "$agent_ctl_binary_name" || true)
+
+	[ -z "$agent_ctl_path" ] && \
+		die "agent control binary does not exist: $agent_ctl_binary_name"
+
+	local redirect="&>\"${ctl_log_file}\""
+
+	if [ "$dry_run" = 'true' ]
+	then
+		redirect=""
+		bundle_dir="FIXME-set-to-OCI-bundle-directory"
+	fi
+
+	run_cmd \
+		sudo \
+		RUST_BACKTRACE=full \
+		"${agent_ctl_path}" \
+		-l debug \
+		connect \
+		--server-address "${server_addr}" \
+		--bundle-dir "${bundle_dir}" \
+		"${cmds}" \
+		$redirect
+}
+
+# This function "cheats" a little - it gets the agent
+# to do some work *and then* stops it.
+stop_local_agent()
+{
+	local cmds=()
+
+	cmds+=("-c Check")
+	cmds+=("-c GetGuestDetails")
+	cmds+=("-c 'sleep 1s'")
+	cmds+=("-c DestroySandbox")
+
+	run_agent_ctl \
+		"${local_agent_ctl_server_addr}" \
+		"${cmds[@]}"
+}
+
+# Doesn't fail. Instead it will return the empty string on error.
+get_agent_vsock_address_simple()
+{
+	local addresses
+
+	addresses=$(ss -HO --vsock |\
+		awk '$2 ~ /^ESTAB$/ {print $6}')
+
+	[ -z "$addresses" ] && return 1
+
+	local expected_count=1
+
+	local count
+	count=$(echo "$addresses"|wc -l || true)
```

Comment:
uhmm, even when the string is empty, wc returns 1 :thinking: 

```
$ echo "" | wc -l
1
```

Reply:
True, but the code has already checked to make sure `$addresses` is not empty.